test/unit/test_vfs2.c

```diff
@@ -385,3 +385,45 @@ TEST(vfs2, rollback, set_up, tear_down, 0, NULL)
 
 	return MUNIT_OK;
 }
+
+/**
+ * Two-node test covering the full replication cycle.
+ */
+TEST(vfs2, leader_and_follower, set_up, tear_down, 0, NULL)
+{
+	struct fixture *f = data;
+	struct node *leader = &f->nodes[0];
+	struct node *follower = &f->nodes[1];
+
+	/* The leader executes and polls a transaction. */
+	sqlite3 *leader_db = open_test_db(leader);
+	OK(sqlite3_exec(leader_db, "CREATE TABLE foo (n INTEGER)", NULL, NULL,
+			NULL));
+	sqlite3_file *leader_fp = main_file(leader_db);
+	dqlite_vfs_frame *frames;
+	struct vfs2_wal_slice leader_sl;
+	OK(vfs2_poll(leader_fp, &frames, &leader_sl));
+	munit_assert_uint(leader_sl.len, ==, 2);
+
+	/* The follower receives the transaction. */
+	sqlite3 *follower_db = open_test_db(follower);
+	sqlite3_file *follower_fp = main_file(follower_db);
+	struct vfs2_wal_slice follower_sl;
+	OK(vfs2_add_uncommitted(follower_fp, PAGE_SIZE, frames, leader_sl.len,
+				&follower_sl));
```

Comment:
Do we need to assert something here about the frames being added? This comment also applies to the unhide and apply below.

We might already be asserting everything that we need to in the invariants themselves but I am not sure.

Reply:
For unhide and apply I think assertions make sense and will add them. For add_uncommitted the only test I can think of that works outside the VFS is to attempt a write transaction and assert that it fails with `SQLITE_BUSY`, indicating that the write lock is held---but that is the kind of thing that the invariant is supposed to cover.

Reply:
We must be updating something for `add_uncommitted`, right? Aren't we changing the wal_cursor? Maybe we are not updating any public field, in that case I am okay not including the check there.

Reply:
Indeed, the wal_cursor is updated. That's not directly observable from outside but I can add an assertion for the size of the WAL on disk.

Reply:
Added more assertions to this test (https://github.com/canonical/dqlite/pull/674/commits/f902cba4ab01c854f0f2129a17263f178921cf19) and caught a bug in the process (https://github.com/canonical/dqlite/pull/674/commits/d924a0674576ae0f14e5f83640eb72ddbb468899)---thanks!